File: recovery/tools/_verify_gate.py
import json, re
from collections import Counter

# 1. Canonical data integrity
d = json.load(open(r'C:\Users\80996\Documents\项目\像素铁道\data\core\railway_data.json', encoding='utf-8'))
print("=== Canonical Data ===")
print(f"Lines: {len(d.get('lineIds', {}))}")
print(f"Stations: {len(d.get('stations', {}))}")
print(f"NameMap: {len(d.get('name_map', {}))}")
print(f"Tourism: {len(d.get('tourism', {}))}")

# 2. Relation Layer
rels = open(r'C:\Users\80996\Documents\项目\像素铁道\data\core\line-service-relations.js', encoding='utf-8').read()
entries = re.findall(r'lineA:\s*"(\w+)"', rels)
types = re.findall(r'relation:\s*"(\w+)"', rels)
print("\n=== Relation Layer ===")
print(f"Entries: {len(entries)}")
print(f"Types: {dict(Counter(types))}")

# 3. Forbidden patterns in resolver
res = open(r'C:\Users\80996\Documents\项目\像素铁道\js\running-chain-resolver.js', encoding='utf-8').read()
print("\n=== Resolver Boundary Check ===")
for p in ['RunningChainDB', 'CHAIN_DB', 'localStorage', 'sessionStorage', 'persistent']:
    status = 'FORBIDDEN!' if p in res else 'OK'
    print(f"  {p}: {status}")

# 4. No delay aggregation in fusion or state
df = open(r'C:\Users\80996\Documents\项目\像素铁道\js\data-fusion.js', encoding='utf-8').read()
ds = open(r'C:\Users\80996\Documents\项目\像素铁道\js\data-state.js', encoding='utf-8').read()
print("\n=== Delay Aggregation Check ===")
print(f"  data-fusion.js: {'WARNING' if 'aggregate' in df.lower() else 'OK - no aggregation'}")
print(f"  data-state.js: {'WARNING' if 'aggregate' in ds.lower() else 'OK - no aggregation'}")

# 5. Check UNKNOWN badge path exists
print("\n=== UNKNOWN Badge Path ===")
print(f"  data-state.js has UNKNOWN branch: {'OK' if 'identity === \"UNKNOWN\"' in ds else 'MISSING!'}")
print(f"  rs-chain-badge-unknown CSS: {'OK' if 'rs-chain-badge-unknown' in open(r'C:\\Users\\80996\\Documents\\项目\\像素铁道\\css\\style.css', encoding='utf-8').read() else 'MISSING!'}")

# 6. Check BRANCH_OF fix
print("\n=== BRANCH_OF Semantic Fix ===")
print(f"  resolver sets SAME for BRANCH: {'OK' if 'ctx.identity=\"SAME\";ctx.confidence=\"LOW\";ctx.reason=\"BRANCH_OF\"' in res else 'STILL WRONG!'}")
print(f"  resolver does NOT set UNKNOWN for BRANCH: {'OK' if 'ctx.identity=\"UNKNOWN\";ctx.confidence=\"LOW\";ctx.reason=\"BRANCH_OF\"' not in res else 'STILL BROKEN!'}")

# 7. Load order in HTML
rt = open(r'C:\Users\80996\Documents\项目\像素铁道\pages\realtime.html', encoding='utf-8').read()
tr = open(r'C:\Users\80996\Documents\项目\像素铁道\pages\trains.html', encoding='utf-8').read()
print("\n=== HTML Load Order ===")
print(f"  realtime.html: {'OK' if 'line-service-relations' in rt and 'running-chain-resolver' in rt and rt.index('running-chain-resolver') < rt.index('data-fusion') else 'WRONG ORDER!'}")
print(f"  trains.html: {'OK' if 'line-service-relations' in tr and 'running-chain-resolver' in tr and tr.index('running-chain-resolver') < tr.index('data-fusion') else 'WRONG ORDER!'}")

print("\n=== SUMMARY ===")
all_ok = (
    len(d.get('lineIds', {})) == 156
    and len(d.get('stations', {})) == 509
    and len(entries) == 14
    and 'aggregate' not in df.lower()
    and 'aggregate' not in ds.lower()
    and 'identity === "UNKNOWN"' in ds
    and 'rs-chain-badge-unknown' in open(r'C:\Users\80996\Documents\项目\像素铁道\css\style.css', encoding='utf-8').read()
    and all(p not in res for p in ['RunningChainDB', 'localStorage'])
    and 'ctx.identity="SAME";ctx.confidence="LOW";ctx.reason="BRANCH_OF"' in res
)
print(f"All checks passed: {all_ok}")

File: recovery/tools/_verify_gate.js
const fs = require('fs');
const ds = fs.readFileSync('js/data-state.js', 'utf8');
const df = fs.readFileSync('js/data-fusion.js', 'utf8');
const css = fs.readFileSync('css/style.css', 'utf8');
const res = fs.readFileSync('js/running-chain-resolver.js', 'utf8');
const rels = fs.readFileSync('data/core/line-service-relations.js', 'utf8');

console.log('=== 4.3.36 Final Verification Gate ===');
console.log('');

// 1. Canonical data
const d = JSON.parse(fs.readFileSync('data/core/railway_data.json', 'utf8'));
console.log('1. Canonical Data:');
console.log('   Lines:', Object.keys(d.lines || {}).length, '(expect 156)');
console.log('   Stations:', Object.keys(d.stations || {}).length, '(expect 509)');
console.log('   NameMap:', Object.keys(d.name_map || {}).length, '(expect 1703)');
console.log('   Tourism:', Object.keys(d.tourism || {}).length, '(expect 93)');

// 2. Relations
const entries = (rels.match(/lineA:/g) || []).length;
const types = rels.match(/relation:\s*"(\w+)"/g) || [];
const tc = {};
types.forEach(t => { tc[t] = (tc[t] || 0) + 1; });
console.log('');
console.log('2. Relation Layer:');
console.log('   Entries:', entries, '(expect 14)');
console.log('   Types:', JSON.stringify(tc));

// 3. Resolver boundary
console.log('');
console.log('3. Resolver Boundary:');
['RunningChainDB','localStorage','sessionStorage','persistent','CHAIN_DB'].forEach(p => {
  console.log('   ' + p + ':', res.includes(p) ? 'FORBIDDEN!' : 'OK');
});

// 4. No aggregation
console.log('');
console.log('4. Delay Aggregation:');
console.log('   data-fusion.js:', df.toLowerCase().includes('aggregate') ? 'WARNING' : 'OK - no aggregation');
console.log('   data-state.js:', ds.toLowerCase().includes('aggregate') ? 'WARNING' : 'OK - no aggregation');

// 5. UNKNOWN badge
console.log('');
console.log('5. UNKNOWN Badge:');
console.log('   data-state.js branch:', ds.includes('identity === "UNKNOWN"') ? 'OK' : 'MISSING!');
console.log('   CSS rule:', css.includes('rs-chain-badge-unknown') ? 'OK' : 'MISSING!');

// 6. BRANCH_OF fix
console.log('');
console.log('6. BRANCH_OF Semantic Fix:');
console.log('   Sets SAME:', res.includes('ctx.identity="SAME";ctx.confidence="LOW";ctx.reason="BRANCH_OF"') ? 'OK' : 'STILL BROKEN!');
console.log('   Does NOT set UNKNOWN:', !res.includes('ctx.identity="UNKNOWN";ctx.confidence="LOW";ctx.reason="BRANCH_OF"') ? 'OK' : 'STILL WRONG!');

// 7. Load order
const rt = fs.readFileSync('pages/realtime.html', 'utf8');
const tr = fs.readFileSync('pages/trains.html', 'utf8');
console.log('');
console.log('7. HTML Load Order:');
console.log('   realtime.html:', rt.indexOf('running-chain-resolver') < rt.indexOf('data-fusion') ? 'OK' : 'WRONG!');
console.log('   trains.html:', tr.indexOf('running-chain-resolver') < tr.indexOf('data-fusion') ? 'OK' : 'WRONG!');

// 8. Summary
const allOk = 
  Object.keys(d.lines || {}).length === 156 &&
  Object.keys(d.stations || {}).length === 509 &&
  entries === 14 &&
  !df.toLowerCase().includes('aggregate') &&
  !ds.toLowerCase().includes('aggregate') &&
  ds.includes('identity === "UNKNOWN"') &&
  css.includes('rs-chain-badge-unknown') &&
  res.includes('ctx.identity="SAME";ctx.confidence="LOW";ctx.reason="BRANCH_OF"') &&
  !res.includes('ctx.identity="UNKNOWN";ctx.confidence="LOW";ctx.reason="BRANCH_OF"') &&
  rt.indexOf('running-chain-resolver') < rt.indexOf('data-fusion') &&
  tr.indexOf('running-chain-resolver') < tr.indexOf('data-fusion') &&
  !res.includes('RunningChainDB') &&
  !res.includes('localStorage');

console.log('');
console.log('=== FINAL: ' + (allOk ? 'ALL CHECKS PASSED' : 'SOME CHECKS FAILED') + ' ===');
